### ISSUE TITLE
Feature/bundle request: Julia

### ISSUE THREAD

**Issue by halars** [clearlinux/distribution#126]
It would be useful to have Julia (https://julialang.org) available on Clear Linux as a bundle. It fits well within the scope of Clear Linux in my opinion.

The pre-compiled "Generic Linux Binaries" available (https://julialang.org/downloads), already run without any issues--that I have noticed--so any extreme work to implement it as a bundle, shouldn't be required.

The benefits would be easy deployment and updates, and potentially some performance optimisations.

**Comment by fenrus75** [comment]
I looked at Julia a year ago.... I’ll look again tomorrow

**Comment by fenrus75** [comment]
so far it is julia 4 - arjan 0

still poking at it

**Comment by iphutch** [comment]
@fenrus75 Any wins to your name in this space?

**Comment by rahulunair** [comment]
@iphutch @fenrus75  any update on this ?

**Comment by chriselrod** [comment]
I'm running into issues trying to build Julia from source on Clear Linux that I've never seen before:

```
gfortran: error: unrecognized command line option ‘-i8’; did you mean ‘-k8’?
```

I've never had problems building from source before on other distros (Ubuntu / Fedora / Arch / Antergos).

`-fdefault-integer-8` is the correct flag for using 8-byte integers.
The errors seem to come up while building OpenBLAS. The [makefile](https://github.com/xianyi/OpenBLAS/blob/89b60dab8ad21a0cc6320cbd9fcd603c4c4bfc81/Makefile.system) checks:
```
ifeq ($(F_COMPILER), GFORTRAN)
CCOMMON_OPT += -DF_INTERFACE_GFORT
FCOMMON_OPT += -Wall
# make single-threaded LAPACK calls thread-safe #1847
FCOMMON_OPT += -frecursive
#Don't include -lgfortran, when NO_LAPACK=1 or lsbcc
ifneq ($(NO_LAPACK), 1)
EXTRALIB += -lgfortran
endif
ifdef NO_BINARY_MODE
ifeq ($(ARCH), $(filter $(ARCH),mips64))
ifdef BINARY64
FCOMMON_OPT += -mabi=64
else
FCOMMON_OPT += -mabi=n32
endif
else ifeq ($(ARCH), $(filter $(ARCH),mips))
FCOMMON_OPT += -mabi=32
endif
else
ifdef BINARY64
ifneq ($(OSNAME), AIX)
FCOMMON_OPT += -m64
endif
ifdef INTERFACE64
ifneq ($(INTERFACE64), 0)
FCOMMON_OPT +=  -fdefault-integer-8
endif
endif
else
ifneq ($(OSNAME), AIX)
FCOMMON_OPT += -m32
endif
endif
endif
ifeq ($(USE_OPENMP), 1)
FCOMMON_OPT += -fopenmp
endif
endif

ifeq ($(F_COMPILER), INTEL)
CCOMMON_OPT += -DF_INTERFACE_INTEL
ifdef INTERFACE64
ifneq ($(INTERFACE64), 0)
FCOMMON_OPT += -i8
endif
endif
ifeq ($(USE_OPENMP), 1)
FCOMMON_OPT += -fopenmp
endif
endif
```

Similarly, Julia's build system also checks:
```
USE_BLAS_FFLAGS :=
ifeq ($(USE_BLAS64), 1)
ifeq ($(USEIFC),1)
USE_BLAS_FFLAGS += -i8
else
USE_BLAS_FFLAGS += -fdefault-integer-8
endif
endif
```

So it seems like, somewhere, some build system is getting confused and accidentally passing '-i8' instead of '-fdefault-integer-8'. Unfortunately, Julia's package ecosystem [assumes](https://github.com/JuliaLinearAlgebra/Arpack.jl/issues/7) Julia has been built with [64-bit OpenBLAS](https://github.com/JuliaLinearAlgebra/Arpack.jl/issues/5). If it weren't for this, I'd build with either MKL or 32-bit OpenBLAS.

Any idea how I may be able to address this? Or why the build systems are confused?

**Comment by iphutch** [comment]
@bryteise

**Comment by bryteise** [comment]
I am sadly not likely to get to this one soon but we definitely want to add it. As to the oddity @chriselrod hit I'd be a little surprised that it is an OpenBLAS bug though that is not a fun package to build. Could be just how julia is trying to build with it? As an aside you can see how Clear Linux is building OpenBLAS via: https://github.com/clearlinux-pkgs/openblas/blob/master/openblas.spec

**Comment by chriselrod** [comment]
Using "INTERFACE64=1" triggers the specific error I was getting, even when cloning OpenBLAS directly and trying to build it. The error was:
`gfortran: error: unrecognized command line option ‘-i8’; did you mean ‘-k8’?`
The error comes up because `ifort` uses `-i8` for 64-bit integers, while `gfortran` uses `-fdefault-integer-8`, and the OpenBLAS build system gets confused into defining `F_COMPILER=Intel`.

The solution I used, which I see Clear Linux's build script was using too, was passing "F_COMPILER=GFORTRAN".

I don't know perl, but I think [this code](https://github.com/xianyi/OpenBLAS/blob/5cf18e28754187800458acca501f7939f6912c22/f_check#L66) from the OpenBLAS build system is to blame. That is, it runs:
`gfortran -O2 -S ftest.f > /dev/null 2>&1 && cat ftest.s && rm -f ftest.s`:
```asm
	.file	"ftest.f"
	.text
	.p2align 4,,15
	.globl	zhoge_
	.type	zhoge_, @function
zhoge_:
.LFB0:
	.cfi_startproc
	pxor	%xmm1, %xmm1
	pxor	%xmm0, %xmm0
	ret
	.cfi_endproc
.LFE0:
	.size	zhoge_, .-zhoge_
	.ident	"GCC: (Clear Linux OS for Intel Architecture) 8.2.1 20190208 gcc-8-branch@268661"
	.section	.note.GNU-stack,"",@progbits
```
and then I think it pattern matches the "Intel" at the end to identify the compiler as ifort. =/
Hence, the need for explicit "F_COMPILER=GFORTRAN".

**Comment by benjaminglass1** [comment]
It should be possible to override Julia's custom build of OpenBLAS and force it to use the system version using two make flags:

``` make
USE_SYSTEM_BLAS = 1
USE_SYSTEM_LAPACK = 1
```

This behavior is documented in Julia's [Make.inc file](https://github.com/JuliaLang/julia/blob/master/Make.inc) (lines 43 and 44) and utilized in the [Arch Linux build](https://git.archlinux.org/svntogit/community.git/tree/trunk/Make.user?h=packages/julia) (lines 22-24) of Julia, for reference.

**Comment by chriselrod** [comment]
Recently I've been building Julia (on Clear Linux) with MKL.
Before that, IIRC, I could build master (but not the 1.1 release) by simply adding `F_COMPILER=GFORTRAN` to [OPENBLAS_BUILD_OPTS](https://github.com/JuliaLang/julia/blob/master/deps/blas.mk#L7).

I don't know much about licensing. Eg, why [Microsoft can ship R with MKL](https://mran.microsoft.com/documents/rro/multithread) but folks at Julia say it violates the GPL.

**Comment by rana** [comment]
Any movement to bundling [Julia](https://julialang.org/)?

**Comment by rana** [comment]
@fredrikekre Any thoughts on unblocking?

**Comment by rana** [comment]
Julia now builds on Clear Linux with [Julia version 1.2.0-rc1.0](https://github.com/JuliaLang/julia/tree/v1.2.0-rc1).

```sh
git clone git://github.com/JuliaLang/julia.git
cd julia
git checkout v1.2.0-rc1
make # or make -j 16
./julia
               _
   _       _ _(_)_     |  Documentation: https://docs.julialang.org
  (_)     | (_) (_)    |
   _ _   _| |_  __ _   |  Type "?" for help, "]?" for Pkg help.
  | | | | | | |/ _` |  |
  | | |_| | | | (_| |  |  Version 1.2.0-rc1.0 (2019-05-30)
 _/ |\__'_|_|_|\__'_|  |  
|__/                   |

julia> exit()
```

@fredrikekre Mentioned upcoming support for Julia on Alpine. So I figured I'd try building with that branch. Turns out it works.

So next steps. 

How do we get an official Julia bundle? MKL support?

**Comment by bryteise** [comment]
Ah lovely, well we need to tweak julia's build system to not download anything which is not always super straight forward. @ahkok did you have some Julia progress that might be unblocked with the new rc?

**Comment by fenrus75** [comment]
julia builds. it just ends up breaking the whole rest of the distro ;-(

**Comment by rana** [comment]
Which issues are breaking the distro? How to mitigate?

**Comment by srgk26** [comment]
Hi! So is there any update on this? It's not hard to install Julia myself, but it would be nice to have Julia in the official repo, especially one that is explicitly built with MKL. As for MKL, it might be worth checking out this though: https://github.com/JuliaComputing/MKL.jl. Though it's not the same as building using MKL from source, it's just 2-steps to switch from OpenBLAS to MKL.

**Comment by wpeguero** [comment]
Hello I would also be very interested in having Julia. Any update in this regard?

**Comment by chriselrod** [comment]
If you want to build Julia from source locally, it should be fine. BinaryBuilder takes care of most binaries needed, i.e. they'll just be downloaded automatically.

If you also want to build dependencies yourself, you'll need to modify the build script to get OpenBLAS to build as described above:
```diff
diff --git a/deps/openblas.mk b/deps/openblas.mk
index 50873c9220..5a50066619 100644
--- a/deps/openblas.mk
+++ b/deps/openblas.mk
@@ -5,7 +5,7 @@ OPENBLAS_GIT_URL := https://github.com/xianyi/OpenBLAS.git
 OPENBLAS_TAR_URL = https://api.github.com/repos/xianyi/OpenBLAS/tarball/$1
 $(eval $(call git-external,openblas,OPENBLAS,,,$(BUILDDIR)))
 
-OPENBLAS_BUILD_OPTS := CC="$(CC)" FC="$(FC)" LD="$(LD)" RANLIB="$(RANLIB)" TARGET=$(OPENBLAS_TARGET_ARCH) BINARY=$(BINARY)
+OPENBLAS_BUILD_OPTS := CC="$(CC)" FC="$(FC)" LD="$(LD)" RANLIB="$(RANLIB)" TARGET=$(OPENBLAS_TARGET_ARCH) BINARY=$(BINARY) F_COMPILER=GFORTRAN
 
 # Thread support
 ifeq ($(OPENBLAS_USE_THREAD), 1)
```
FWIW, if you have an Intel CPU and care about BLAS/LAPACK performance, I'd generally recommend using MKL.jl (e.g., I have `using MKL` in my `~/.julia/config/startup.jl`) and not worry too much about OpenBLAS.

It is not recommended to use Julia with system libraries. That is important to consider for anyone trying to package Julia with their distro.
It's much better to just download an official binary or build from source; Julias installed from distro repos are often broken in some way.

Also worth pointing out that Clear Linux is still on LLVM 11, while Julia's latest release is on LLVM 12.
